src/OpenDefinitions.elm

```diff
@@ -1,165 +1,341 @@
+{-
+   OpenDefinitions is the main data structure for working with definitions in
+   the Workspace area. It supports holding Loading definitions and inserting
+   definitions after another.
+
+   It features a `focus` indicator and a before and after just like a Zipper.
+   `focus` can be changed with `next` and `prev`.
+
+   Invariants:
+     It structurally can't hold the invariant that it should not contain
+     duplicates, so this is instead enforced via the API; uniqueness is
+     determined by Hash.
+-}
+
+
 module OpenDefinitions exposing
-    ( OpenDefinitions
-    , definitions
+    ( HashIndexedDefinition
+    , OpenDefinitions
     , empty
-    , fromList
-    , get
+    , focus
+    , fromDefinitions
     , hashes
-    , insert
-    , insertAfter
-    , loading
+    , init
+    , insertWithFocus
+    , insertWithFocusAfter
+    , isEmpty
+    , isFocused
+    , map
+    , mapToList
     , member
+    , next
+    , prev
     , remove
     , replace
-    , replaceItems
+    , singleton
     , toList
-    , update
     )
 
 import Definition exposing (Definition)
 import Hash exposing (Hash)
-import OrderedDict exposing (OrderedDict)
+import List
+import List.Extra as ListE
 import RemoteData exposing (WebData)
 
 
-
-{-
-   OpenDefinitions is the main data structure for working with definitions in
-   the Workspace area. It wraps an OrderedDirect (Hash to Definition), but is
-   opaque to that underlying data structure and only exposes a few key APIs,
-   including a few convenience functions for working with, loading and opening
-   Definitions.
+{-| This technically allows multiple of the same definition across the 3 fields.
+This is conceptionally not allowed and is enforced by the helper functions.
 -}
+type OpenDefinitions
+    = Empty
+    | OpenDefinitions
+        { before : List HashIndexedDefinition
+        , focus : HashIndexedDefinition
+        , after : List HashIndexedDefinition
```

Comment:
Another representation you could consider is just `{ defs : List HashIndexedDefinition, focus : Int }`. I'm not sure the zipper-like representation you have is buying you all that much, especially since you already need random access of the focus to support the user clicking on any row in the UI to focus on it. 

Even though I think zippers are a cool idea, the way I do things like that is almost always just a regular data structure paired with a path or index into that structure. Some advantages beyond simplicity are you get to use the ordinary functions of that data structure, rather than needing to write things that are zipper specific (for instance, mapping over the data structure is easy, reversing it, taking the first 10 elements of it, etc, are all easier when you have the regular data structure rather than a zipper into it). Random access is more naturally supported - you can jump the focus around directly without going through intermediate states. And you can straightforwardly have a notion of multiple focuses if you want.

You can also use `Array HashIndexedDefinition` if the list is big and you want faster random access, but I don't think that matters here.

Reply:
I'm preferring the Zipper approach mostly to be able to communicate the invariants of this data structure and to be more or less forced to consider the focus when modifying the contents with things like `remove`. There's definitely something to be said about simplicity and access to existing APIs and I could likely make this simpler while keeping it a Zipper by just using a Zipper library.

Regarding Array; It's certainly good to keep in mind, but I don't think we'll hit the limits of List in most parts of the UI, least of all the open definitions. That said, it might eventually be needed for things like the namespace tree; the top level namespace tree for the "hub" version of the UI will (hopefully) become a very large container :)